proxylib/binarymemcached/parser.go

```
@@ -0,0 +1,341 @@
+	"github.com/cilium/cilium/proxylib/proxylib"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// BinaryMemcacheRule matches against memcached requests
+type BinaryMemcacheRule struct {
+	// opCode group name
+	opCode string
+
+	keyExact  string
+	keyPrefix string
+	keyRegex  string
+
+	// allowed opcodes
+	opCodes []byte
+	// compiled regex
+	regex *regexp.Regexp
+}
+
+type binaryMemcacheMeta struct {
+	opCode byte
+	key    string
+}
+
+// Matches returns true if the BinaryMemcacheRule matches
+func (rule *BinaryMemcacheRule) Matches(data interface{}) bool {
+	log.Infof("binarymemcache checking rule %v", *rule)
+
+	packetMeta := data.(binaryMemcacheMeta)
+
+	if rule.keyExact != "" {
+		return rule.keyExact == packetMeta.key && rule.matchOpcode(packetMeta.opCode)
+	}
+
+	if rule.keyPrefix != "" {
+		return strings.HasPrefix(packetMeta.key, rule.keyPrefix) && rule.matchOpcode(packetMeta.opCode)
+	}
+
+	if rule.keyRegex != "" {
+		return rule.matchOpcode(packetMeta.opCode) && rule.regex.MatchString(packetMeta.key)
+	}
+
+	log.Infof("No key rule specified, matching by opcode")
+
+	return rule.matchOpcode(packetMeta.opCode)
+}
+
+func (rule *BinaryMemcacheRule) matchOpcode(oc byte) bool {
+	return bytes.Contains(rule.opCodes, []byte{oc})
+}
+
+func getMemcacheKey(packet []byte, extrasLength, keyLength uint32) string {
+	return string(packet[headerSize+extrasLength : headerSize+(extrasLength)+keyLength])
+}
+
+// L7BinaryMemcacheRuleParser parses protobuf L7 rules to and array of BinaryMemcacheRule
+// May panic
+func L7BinaryMemcacheRuleParser(rule *cilium.PortNetworkPolicyRule) []proxylib.L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		panic(fmt.Errorf("Can't get L7 rules"))
+	}
+	var rules []proxylib.L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var br BinaryMemcacheRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "opCode":
+				br.opCode = v
+				br.opCodes = MemcacheOpCodeMap[v]
```

Is opcode mandatory part of the rule? It is matched unconditionally, so maybe should panic after this for loop if opcode is not specified?

---

Good idea, will do.